src/jit/gentree.cpp

```diff
@@ -12568,38 +12568,58 @@ GenTree* Compiler::gtOptimizeEnumHasFlag(GenTree* thisOp, GenTree* flagOp)
     // Yes, both boxes can be cleaned up. Optimize.
     JITDUMP("Optimizing call to Enum.HasFlag\n");
 
-    // Pop the stack for real.
-    // impPopStack();
-    // impPopStack();
-
-    // Undo the boxing and prepare to operate directly
+    // Undo the boxing of thisOp and prepare to operate directly
     // on the original enum values.
     thisVal = gtTryRemoveBoxUpstreamEffects(thisOp, BR_REMOVE_BUT_NOT_NARROW);
-    flagVal = gtTryRemoveBoxUpstreamEffects(flagOp, BR_REMOVE_BUT_NOT_NARROW);
 
     // Our trial removal above should guarantee successful removal here.
     assert(thisVal != nullptr);
-    assert(flagVal != nullptr);
-
-    // The thisVal and flagVal trees come from earlier statements. We need
-    // to evaluate them both to temps at those points to safely transmit
-    // the values here.
-    const unsigned flagTmp         = lvaGrabTemp(true DEBUGARG("Enum:HasFlag flag temp"));
-    GenTree*       flagAsg         = gtNewTempAssign(flagTmp, flagVal);
-    GenTree*       flagAsgStmt     = flagOp->AsBox()->gtCopyStmtWhenInlinedBoxValue;
-    flagAsgStmt->gtStmt.gtStmtExpr = flagAsg;
-
-    const unsigned thisTmp         = lvaGrabTemp(true DEBUGARG("Enum:HasFlag this temp"));
-    GenTree*       thisAsg         = gtNewTempAssign(thisTmp, thisVal);
-    GenTree*       thisAsgStmt     = thisOp->AsBox()->gtCopyStmtWhenInlinedBoxValue;
-    thisAsgStmt->gtStmt.gtStmtExpr = thisAsg;
-
-    // Turn the call into (thisTmp & flagTmp) == flagTmp.
+
+    // We should have a consistent view of the type
     var_types type = thisVal->TypeGet();
     assert(type == flagVal->TypeGet());
 
-    GenTree* andTree = gtNewOperNode(GT_AND, type, gtNewLclvNode(thisTmp, type), gtNewLclvNode(flagTmp, type));
-    GenTree* cmpTree = gtNewOperNode(GT_EQ, type, andTree, gtNewLclvNode(flagTmp, type));
+    // The thisVal and flagVal trees come from earlier statements.
+    //
+    // Unless they are invariant values, we need to evaluate them both
+    // to temps at those points to safely transmit the values here.
+    //
+    // Also we need ot use the flag twice, so we need two trees for it.
```

Comment:
typo: "ot" -> "to"

Reply:
Found one other typo; will fix them both.